File: laravel/app/Models/Players/Achievement.php
<?php

namespace App\Models\Players;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Database\Eloquent\Relations\BelongsToMany;

class Achievement extends Model
{
    protected $table = 'achievements';

    protected $fillable = ['name', 'description', 'icon_url', 'points', 'rarity', 'is_hidden'];

    protected $casts = [
        'is_hidden' => 'boolean',
    ];

    const RARITY_VALUES = ['Common', 'Uncommon', 'Rare', 'Epic', 'Legendary'];

    // ── Validation rules ─────────────────────────────────────────────
    public function validateRules(): void
    {
        $errors = [];
        if (!(($this->points === null || $this->points > 0))) {
            $errors['points_positive'] = 'Achievement must award at least one point';
        }
        if (!empty($errors)) {
            throw new \Illuminate\Validation\ValidationException(
                \Illuminate\Support\Facades\Validator::make([], []),
                response()->json(['errors' => $errors], 422)
            );
        }
    }

    // ── Business operations ──────────────────────────────────────────

    public function pointValue($multiplier): int
    {
        throw new \RuntimeException('point_value not implemented');
    }

    public function reveal(): void
    {
        throw new \RuntimeException('reveal not implemented');
    }

}

File: phoenix/lib/cards_project/players/achievement.ex
defmodule CardsProject.Players.Achievement do
  use Ecto.Schema
  import Ecto.Changeset

  schema "achievements" do
    field :name, :string
    field :description, :string
    field :icon_url, :string
    field :points, :integer, default: 10
    field :rarity, :string
    field :is_hidden, :boolean, default: false
    many_to_many :players, CardsProject.Players.Player, join_through: "player_achievements"
    has_many :player_records, CardsProject.Players.PlayerAchievement, foreign_key: :achievement_id

    timestamps()
  end

  @doc false
  def changeset(record, attrs) do
    record
    |> cast(attrs, [:name, :description, :points, :is_hidden, :icon_url, :rarity])
    |> validate_required([:name, :description, :points, :is_hidden])
    |> validate_inclusion(:rarity, ["Common", "Uncommon", "Rare", "Epic", "Legendary"])
    |> validate_number(:points, greater_than: 0, message: "Achievement must award at least one point")
  end

  # ── Business operations ────────────────────────────────────────────

  def point_value(_record, _multiplier) do
    # TODO: implement Achievement.point_value
    {:error, :not_implemented}
  end

  def reveal(_record) do
    # TODO: implement Achievement.reveal
    :ok
  end
end
